// graphql-explorer/lib/ConnectionSection.js
"use strict";

exports.__esModule = true;
exports.default = ConnectionSection;

var _react = _interopRequireWildcard(require("react"));

var _Button = _interopRequireDefault(require("react-bootstrap/Button"));

var _Spinner = _interopRequireDefault(require("react-bootstrap/Spinner"));

var _ListSection = require("./ListSection");

var _PanelContainer = _interopRequireDefault(require("./ui/PanelContainer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ConnectionSection({
  item,
  itemType,
  executeQuery,
  input
}) {
  const [connection, setconnection] = (0, _react.useState)(item);
  const [loading, setLoading] = (0, _react.useState)(false);
  const items = connection.edges.map(({
    node
  }) => node);
  const handleMore = (0, _react.useCallback)(async () => {
    setLoading(true);
    const resp = await executeQuery(undefined, Object.assign({}, input, {
      after: connection.pageInfo.endCursor
    }));
    setconnection({
      edges: [...connection.edges, ...resp.edges],
      pageInfo: resp.pageInfo
    });
    setLoading(false);
  }, [connection.edges, connection.pageInfo.endCursor, executeQuery, input]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ListSection.ListSectionBase, {
    items: items,
    itemType: itemType
  }), /*#__PURE__*/_react.default.createElement(_PanelContainer.default.Body, null, /*#__PURE__*/_react.default.createElement(_Button.default, {
    block: true,
    onClick: handleMore,
    disabled: !executeQuery || loading || !connection.pageInfo.hasNextPage
  }, loading ? /*#__PURE__*/_react.default.createElement(_Spinner.default, {
    size: "sm",
    animation: "border"
  }) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "+"))));
}
